src/rust/src/file_reader.rs

```diff
@@ -465,29 +486,46 @@ impl<R: Read, S> FeatureIter<R, S> {
     }
 
     fn advance_finished(&mut self) -> bool {
-        if self.finished {
+        if self.state == State::Finished {
             return true;
         }
         if let Some(count) = self.count {
             if self.feat_no >= count {
-                self.finished = true;
+                self.state = State::Finished;
                 return true;
             }
         }
         false
     }
 
     /// Read feature size and return true if end of dataset reached
-    fn read_feature_size(fbs: &mut FgbFeature, reader: &mut R) -> bool {
-        fbs.feature_buf.resize(4, 0);
-        reader.read_exact(&mut fbs.feature_buf).is_err()
+    fn read_feature_size(&mut self) -> bool {
+        self.fbs.feature_buf.resize(4, 0);
+        self.cur_pos += 4;
+        self.reader.read_exact(&mut self.fbs.feature_buf).is_err()
     }
 
     fn read_feature(&mut self) -> Result<()> {
-        // Read feature size if not already read in select_all or select_bbox
-        if self.cur_pos != 4 && Self::read_feature_size(&mut self.fbs, &mut self.reader) {
```

Comment:
`self.cur_pos != 4` was our sentinel for "This is the first feature and we've already called read_feature_size in the initializer so don't do it again" - now we track an explicit State variable, which I think is a little more explicable.